### ISSUE TITLE
HTTP Header Proxy-Authenticate can be conditionally forwarded.

### ISSUE THREAD

**Issue by dakriy** [traefik/traefik#7374]
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?

When using the forwardAuth middleware and a request using the `Proxy-Authenticate` header, I expect the header to be forwarded to the authentication server.

### Justification

Let me start by saying I am aware that Proxy-Authenticate is a hop by hop header and should normally be stripped according to [RFC 2616](https://tools.ietf.org/html/rfc2616#page-92). However, [RFC 7235](https://tools.ietf.org/html/rfc7235#section-4.4) extends on authentication headers. In section 4.4 it states:
```
   Unlike Authorization, the Proxy-Authorization header field applies
   only to the next inbound proxy that demanded authentication using the
   Proxy-Authenticate field.  When multiple proxies are used in a chain,
   the Proxy-Authorization header field is consumed by the first inbound
   proxy that was expecting to receive credentials.  A proxy MAY relay
   the credentials from the client request to the next proxy if that is
   the mechanism by which the proxies cooperatively authenticate a given
   request.
```

Since the forward auth middleware is delegating authentication for the proxy, the `Proxy-Authenticate` header can be sent along freeing up the `Authorization` header for applications behind the proxy that also want authentication. This would be Traefik's way of consuming the header. By optionally allowing the header to be passed, it would also allow for proxy chaining with proxy authentication and would also not not hog the application `Authorization` header.

**Comment by Scapal** [comment]
I'm bumping into the same problem when using Traefik forwardAuth with Authelia.

That header is stripped by calling `utils.RemoveHeaders(forwardReq.Header, forward.HopHeaders...)`.
```
var HopHeaders = []string{
	Connection,
	KeepAlive,
	ProxyAuthenticate,
	ProxyAuthorization,
	Te, // canonicalized version of "TE"
	Trailers,
	TransferEncoding,
	Upgrade,
}
```
Since the forward Auth is not considered a next hop but is actually part of the auth process of the current hop, the Proxy-Authorization header should not be removed.

https://github.com/traefik/traefik/blob/7928e6d0cd4c7751a14bea5f74124f8a3cb829c0/pkg/middlewares/auth/forward.go#L163

**Comment by Scapal** [comment]
@nkonev I saw your commit for `authRequestHeaders`.

Could you do the removal of HopHeaders in filterForwardRequestHeaders so it allows to keep `Proxy-Authentication` if listed in `authRequestHeaders` ?

``` go
func filterForwardRequestHeaders(forwardRequestHeaders http.Header, allowedHeaders []string) http.Header {
	if len(allowedHeaders) == 0 {
		return utils.RemoveHeaders(forwardRequestHeaders, forward.HopHeaders...)
	}
```

A better approach would also be to not consider `Proxy-Authentication` as a Hop Header for the Authentication since it is actually supposed to perform that authentication.

``` go
func keepProxyAuthHeader(hopHeaders []string) []string {
	var headers []string
	for _, h := range hopHeaders {
		if h != forward.ProxyAuthorization {
			headers = append(headers, h)
		}
	}
	return headers
}
```
``` go
func writeHeader(req, forwardReq *http.Request, trustForwardHeader bool, allowedHeaders []string) {
	utils.CopyHeaders(forwardReq.Header, req.Header)
       // utils.RemoveHeaders(forwardReq.Header, forward.HopHeaders...)
	forwardReq.Header = filterForwardRequestHeaders(forwardReq.Header, allowedHeaders)
```

``` go
func filterForwardRequestHeaders(forwardRequestHeaders http.Header, allowedHeaders []string) http.Header {
	if len(allowedHeaders) == 0 {
		return utils.RemoveHeaders(forwardReq.Header, keepProxyAuthHeader(forward.HopHeaders)...)
	}
```

**Comment by nkonev** [comment]
@Scapal (Unfortunately) I am not a part of traefik team, I just [solved](https://github.com/traefik/traefik/pull/7226) my problem with an [excess](https://github.com/traefik/traefik/issues/7225) Content-Type header.

I think you should make a PR with change of hop headers.

**Comment by traefiker** [comment]
Closed by #7433.
